sdk/python/kfp/_local_client.py

```diff
@@ -0,0 +1,422 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import datetime
+import json
+import logging
+import re
+import subprocess
+from typing import Any
+from typing import Callable
+from typing import Dict
+from typing import List
+from typing import Mapping
+from typing import cast
+
+import kfp
+
+
+class _Dag:
+    def __init__(self, nodes):
+        self.graph = {node: [] for node in nodes}
+        self.reverse_graph = {node: [] for node in nodes}
+
+    def add_edge(self, edge_from, edge_to):
+        self.graph[edge_from].append(edge_to)
+        self.reverse_graph[edge_to].append(edge_from)
+
+    def get_follows(self, node):
+        return self.graph.get(node, [])
+
+    def get_dependencies(self, node):
+        return self.reverse_graph.get(node, [])
+
+    def topological_sort(self):
+        """ List DAG nodes in topological order. """
+
+        in_degree = {node: 0 for node in self.graph.keys()}
+
+        for i in self.graph:
+            for j in self.graph[i]:
+                in_degree[j] += 1
+
+        queue = []
+        for node, degree in in_degree.items():
+            if degree == 0:
+                queue.append(node)
+
+        top_order = []
+
+        while queue:
+            u = queue.pop(0)
+            top_order.append(u)
+
+            for node in self.graph[u]:
+                in_degree[node] -= 1
+
+                if in_degree[node] == 0:
+                    queue.append(node)
+
+        return top_order
+
+
+class LocalClient:
+    def _extract_pipeline_param(self, param: str):
+        matches = re.findall(
+            r"{{pipelineparam:op=([\w\s_-]*);name=([\w\s_-]+)}}", param
+        )
+        op_dependency_name = matches[0][0]
+        output_file_name = matches[0][1]
+        return kfp.dsl.PipelineParam(output_file_name, op_dependency_name)
+
+    def _pair_arguments(self, cmd: List):
```

Comment:
```suggestion
    def _pair_arguments(self, cmd: List[str]):
```

Reply:
My personal thought is that `_get_keyword_arguments` could be a better name, up to you.